src/cavity.jl

```diff
@@ -0,0 +1,108 @@
+@kwdef mutable struct RFParams{T<:Number} <: AbstractParams
+    frequency::T          = Float32(0.0) # RF frequency in Hz or Harmonic number
+    voltage::T            = Float32(0.0) # Voltage in V 
+    phi0::T               = Float32(0.0) # Phase at reference energy
+    const harmon_master::Bool = false    # false = frequency in Hz, true = harmonic number
+
+    function RFParams(args...)
+        return new{promote_type(map(x->typeof(x),args)...)}(args...)
+    end
+end
+
+Base.eltype(::RFParams{T}) where {T} = T
+Base.eltype(::Type{RFParams{T}}) where {T} = T
+
+
+function Base.isapprox(a::RFParams, b::RFParams) 
+    return  a.frequency     ≈  b.frequency && 
+            a.voltage       ≈  b.voltage && 
+            a.phi0          ≈  b.phi0 && 
+            a.harmon_master == b.harmon_master
+end
+
+# Frequency mapping system analogous to BMULTIPOLE_STRENGTH_MAP
+const CAVITY_FREQUENCY_MAP = Dict{Symbol,Bool}(
+  :rf_frequency    => false,  # harmon_master = false (frequency in Hz)
+  :harmonic_number => true,   # harmon_master = true (harmonic number)
+)
+
+const CAVITY_FREQUENCY_INVERSE_MAP = Dict(value => key for (key, value) in CAVITY_FREQUENCY_MAP)
+
+function Base.hasproperty(c::RFParams, key::Symbol)
+  if key in fieldnames(RFParams)
+    return true
+  elseif haskey(CAVITY_FREQUENCY_MAP, key)
+    harmon_master = CAVITY_FREQUENCY_MAP[key]
+    return harmon_master == c.harmon_master
+  else
+    return false
+  end
+end
+
+function Base.getproperty(c::RFParams, key::Symbol)
+  if key in fieldnames(RFParams)
+    return getfield(c, key)
+  elseif haskey(CAVITY_FREQUENCY_MAP, key)
+    harmon_master = CAVITY_FREQUENCY_MAP[key]
+    if harmon_master == c.harmon_master
+      return c.frequency
+    else
+      correctkey = CAVITY_FREQUENCY_INVERSE_MAP[c.harmon_master]
+      error("Only $correctkey is currently set in RFParams, $key cannot be calculated without particle species information")
+    end
+  end
+  error("RFParams does not have property $key")
+end
+
+function Base.setproperty!(c::RFParams{T}, key::Symbol, value) where {T}
+  if key in (:frequency, :voltage, :phi0)
+    return setfield!(c, key, T(value))
+  elseif key == :harmon_master
+    error("Cannot set `harmon_master` in RFParams directly, set as element parameter instead")
```

Comment:
```suggestion
    return setfield(c, key, value)
```

The `const` part of this field makes this clear